urlaa051fbd3d_WheezePuppet_collegeDiversity_tgz-pJ8-edu_umw_cpsc_collegediversity_SimJ8/src/edu/umw/cpsc/collegediversity/FreshmanHousingSelection.java

```diff
@@ -0,0 +1,214 @@
+package edu.umw.cpsc.collegediversity;
+
+import java.util.*;
+import sim.util.*;
+import ec.util.*;
+
+/**
+ * A Singleton class that can {@link #assign} a set of freshman {@link
+ * Student}s to their freshman dorms. 
+ * @see UpperclassHousingSelection
+ */
+public class FreshmanHousingSelection{
+
+    private static FreshmanHousingSelection theInstance;
+    private Bag dorms;
+
+    private FreshmanHousingSelection(){
+
+        dorms = new Bag();
+
+        Dorm Alvey = new Dorm("Alvey", false, true, 73);
+        Dorm Bushnell = new Dorm("Bushnell", false, true, 76);
+        Dorm Jefferson = new Dorm("Jefferson", false, true, 96);
+        Dorm Randolph = new Dorm("Randolph", false, true, 93);
+        Dorm Russell = new Dorm("Russell", false, true, 87);
+
+        //(actually mixed-year but for simplicity I'm keeping it freshman 
+        // only for now)
+        Dorm Virginia = new Dorm("Virginia", true, true, 92); 
+
+        dorms.add(Virginia);
+        dorms.add(Alvey);
+        dorms.add(Bushnell);
+        dorms.add(Jefferson);
+        dorms.add(Randolph);
+        dorms.add(Russell);
+    }
+
+    /**
+     * Singleton pattern.
+     */
+    public static synchronized FreshmanHousingSelection instance(){
+        if(theInstance==null){
+            theInstance=new FreshmanHousingSelection();
+        }
+        return theInstance;
+    }
+
+    /** 
+     * Assign the set of (presumably freshman, but this is not checked)
+     * students passed to their freshman dorm rooms, in a way that does not
+     * take race into account. (compare {@link #assignByRace}.) As a precursor
+     * (side effect) to this, any freshmen already existing in freshmen dorms
+     * will be removed.
+     */
+    public void assign(Bag newFreshmen) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 84-89 into a separate method.